README.md
# xprezzo-raw-body

Gets the entire buffer of a stream either as a `Buffer` or a string.
Validates the stream's length against an expected length and maximum limit.
Ideal for parsing request bodies.


## Install

This is a [Node.js](https://nodejs.org/en/) module available through the
[npm registry](https://www.npmjs.com/). Installation is done using the
[`npm install` command](https://docs.npmjs.com/getting-started/installing-npm-packages-locally):

```sh
$ npm install xprezzo-raw-body
```

## API

<!-- eslint-disable no-unused-vars -->

```js
var getRawBody = require('xprezzo-raw-body')
```

### getRawBody(stream, [options], [callback])

**Returns a promise if no callback specified and global `Promise` exists.**

Options:

- `length` - The length of the stream.
  If the contents of the stream do not add up to this length,
  an `400` error code is returned.
- `limit` - The byte limit of the body.
  This is the number of bytes or any string format supported by
  [bytes](https://www.npmjs.com/package/bytes),
  for example `1000`, `'500kb'` or `'3mb'`.
  If the body ends up being larger than this limit,
  a `413` error code is returned.
- `encoding` - The encoding to use to decode the body into a string.
  By default, a `Buffer` instance will be returned when no encoding is specified.
  Most likely, you want `utf-8`, so setting `encoding` to `true` will decode as `utf-8`.
  You can use any type of encoding supported by [xprezzo-iconv](https://www.npmjs.org/package/xprezzo-iconv#readme).

You can also pass a string in place of options to just specify the encoding.

If an error occurs, the stream will be paused, everything unpiped,
and you are responsible for correctly disposing the stream.
For HTTP requests, no handling is required if you send a response.
For streams that use file descriptors, you should `stream.destroy()` or `stream.close()` to prevent leaks.

## Errors

This module creates errors depending on the error condition during reading.
The error may be an error from the underlying Node.js implementation, but is
otherwise an error created by this module, which has the following attributes:

  * `limit` - the limit in bytes
  * `length` and `expected` - the expected length of the stream
  * `received` - the received bytes
  * `encoding` - the invalid encoding
  * `status` and `statusCode` - the corresponding status code for the error
  * `type` - the error type

### Types

The errors from this module have a `type` property which allows for the progamatic
determination of the type of error returned.

#### encoding.unsupported

This error will occur when the `encoding` option is specified, but the value does
not map to an encoding supported by the [iconv-lite](https://www.npmjs.org/package/xprezzo-iconv#readme)
module.

#### entity.too.large

This error will occur when the `limit` option is specified, but the stream has
an entity that is larger.

#### request.aborted

This error will occur when the request stream is aborted by the client before
reading the body has finished.

#### request.size.invalid

This error will occur when the `length` option is specified, but the stream has
emitted more bytes.

#### stream.encoding.set

This error will occur when the given stream has an encoding set on it, making it
a decoded stream. The stream should not have an encoding set and is expected to
emit `Buffer` objects.

## Examples

### Simple Express example

```js
var contentType = require('content-type')
var xprezzo = require('xprezzo')
var getRawBody = require('xprezzo-raw-body')

var app = xprezzo()

app.use(function (req, res, next) {
  getRawBody(req, {
    length: req.headers['content-length'],
    limit: '1mb',
    encoding: contentType.parse(req).parameters.charset
  }, function (err, string) {
    if (err) return next(err)
    req.text = string
    next()
  })
})

// now access req.text
```

### Simple Koa example

```js
var contentType = require('content-type')
var getRawBody = require('xprezzo-raw-body')
var koa = require('koa')

var app = koa()

app.use(function * (next) {
  this.text = yield getRawBody(this.req, {
    length: this.req.headers['content-length'],
    limit: '1mb',
    encoding: contentType.parse(this.req).parameters.charset
  })
  yield next
})

// now access this.text
```

### Using as a promise

To use this library as a promise, simply omit the `callback` and a promise is
returned, provided that a global `Promise` is defined.

```js
var getRawBody = require('xprezzo-raw-body')
var http = require('http')

var server = http.createServer(function (req, res) {
  getRawBody(req)
    .then(function (buf) {
      res.statusCode = 200
      res.end(buf.length + ' bytes submitted')
    })
    .catch(function (err) {
      res.statusCode = 500
      res.end(err.message)
    })
})

server.listen(3000)
```

## People

Xprezzo and related projects are maintained by [Cloudgen Wong](mailto:cloudgen.wong@gmail.com).

## License

[MIT](LICENSE)


HISTORY.md
1.1.2 / 2022-05-12
==================

  * depds updated
  * depds xprezzo-bytes updated to 1.1.0
  * depds xprezzo-debug updated to 1.1.2
  * depds xprezzo-iconv updated to 1.1.2
  * depds xprezzo-http-errors updated to 1.1.2
  * depds xprezzo-on-finished updated to 1.1.2
  * depds xprezzo-stream-unpipe updated to 1.1.0
  * depds xprezzo-buffer updated to 1.1.2
  * depds eslint updated to 8.15.0
  * depds eslint-config-standard updated to 17.0.0
  * depds eslint-plugin-import updated to 2.26.0
  * depds eslint-plugin-markdown updated to 2.2.1
  * depds eslint-plugin-promise updated to 6.0.0
  * depds mocha updated to 10.0.0
  * remove eslint-plugin-standard

1.1.1 / 2020-10-19
==================

  * depds xprezzo-debug updated to v1.1.1
  * depds xprezzo-http-errors updated to v1.1.1
  * depds xprezzo-iconv updated to v1.1.1
  * depds xprezzo-on-finished updated to v1.1.1
  * use xprezzo-bytes v1.0.0

1.1.0 / 2020-10-05
==================

  * depds updated

1.0.0 / 2020-10-05
==================

  * Initial release


test/http.js
const assert = require('assert')
const getRawBody = require('..')
const http = require('http')
const net = require('net')

describe('using http streams', () => {
  it('should read body streams', (done) => {
    const server = http.createServer((req, res) => {
      getRawBody(req, { length: req.headers['content-length'] }, (err, body) => {
        if (err) {
          req.resume()
          res.statusCode = 500
          return res.end(err.message)
        }

        res.end(body)
      })
    })

    server.listen(() => {
      const addr = server.address()
      const client = http.request({ method: 'POST', port: addr.port })

      client.end('hello, world!')

      client.on('response', (res) => {
        getRawBody(res, { encoding: true }, (err, str) => {
          server.close(() => {
            assert.ifError(err)
            assert.strictEqual(str, 'hello, world!')
            done()
          })
        })
      })
    })
  })

  it('should throw if stream encoding is set', (done) => {
    const server = http.createServer((req, res) => {
      req.setEncoding('utf8')
      getRawBody(req, { length: req.headers['content-length'] }, (err, body) => {
        if (err) {
          req.resume()
          res.statusCode = 500
          return res.end(err.message)
        }

        res.end(body)
      })
    })

    server.listen(() => {
      const addr = server.address()
      const client = http.request({ method: 'POST', port: addr.port })

      client.end('hello, world!')

      client.on('response', (res) => {
        getRawBody(res, { encoding: true }, (err, str) => {
          server.close(() => {
            assert.ifError(err)
            assert.strictEqual(str, 'stream encoding should not be set')
            done()
          })
        })
      })
    })
  })

  it('should throw if connection ends', (done) => {
    let socket
    const server = http.createServer((req, res) => {
      getRawBody(req, { length: req.headers['content-length'] }, (err, body) => {
        server.close()
        assert.ok(err)
        assert.strictEqual(err.code, 'ECONNABORTED')
        assert.strictEqual(err.expected, 50)
        assert.strictEqual(err.message, 'request aborted')
        assert.strictEqual(err.received, 10)
        assert.strictEqual(err.status, 400)
        assert.strictEqual(err.type, 'request.aborted')
        done()
      })

      setTimeout(socket.destroy.bind(socket), 10)
    })

    server.listen(() => {
      socket = net.connect(server.address().port, () => {
        socket.write('POST / HTTP/1.0\r\n')
        socket.write('Connection: keep-alive\r\n')
        socket.write('Content-Length: 50\r\n')
        socket.write('\r\n')
        socket.write('testing...')
      })
    })
  })
})


test/http2.js
const assert = require('assert')
const getRawBody = require('..')
const tryRequire = (module) => {
  try {
    return require(module)
  } catch (e) {
    return undefined
  }
}
const http2 = tryRequire('http2')
const net = require('net')

const describeHttp2 = !http2
  ? describe.skip
  : describe

describeHttp2('using http2 streams', () => {
  it('should read body streams', (done) => {
    const server = http2.createServer((req, res) => {
      getRawBody(req, { length: req.headers['content-length'] }, (err, body) => {
        if (err) {
          req.resume()
          res.statusCode = 500
          return res.end(err.message)
        }

        res.end(body)
      })
    })

    server.listen(() => {
      const addr = server.address()
      const session = http2.connect('http://localhost:' + addr.port)
      const request = session.request({ ':method': 'POST', ':path': '/' })

      request.end('hello, world!')

      request.on('response', (headers) => {
        getRawBody(request, { encoding: true }, (err, str) => {
          http2close(server, session, () => {
            assert.ifError(err)
            assert.strictEqual(str, 'hello, world!')
            done()
          })
        })
      })
    })
  })

  it('should throw if stream encoding is set', (done) => {
    const server = http2.createServer((req, res) => {
      req.setEncoding('utf8')
      getRawBody(req, { length: req.headers['content-length'] }, (err, body) => {
        if (err) {
          req.resume()
          res.statusCode = 500
          return res.end(err.message)
        }

        res.end(body)
      })
    })

    server.listen(() => {
      const addr = server.address()
      const session = http2.connect('http://localhost:' + addr.port)
      const request = session.request({ ':method': 'POST', ':path': '/' })

      request.end('hello, world!')

      request.on('response', (res) => {
        getRawBody(request, { encoding: true }, (err, str) => {
          http2close(server, session, () => {
            assert.ifError(err)
            assert.strictEqual(str, 'stream encoding should not be set')
            done()
          })
        })
      })
    })
  })

  it('should throw if connection ends', (done) => {
    let socket
    const server = http2.createServer((req, res) => {
      getRawBody(req, { length: req.headers['content-length'] }, (err, body) => {
        server.close()
        assert.ok(err)
        assert.strictEqual(err.code, 'ECONNABORTED')
        assert.strictEqual(err.expected, 50)
        assert.strictEqual(err.message, 'request aborted')
        assert.strictEqual(err.received, 10)
        assert.strictEqual(err.status, 400)
        assert.strictEqual(err.type, 'request.aborted')
        done()
      })

      setTimeout(socket.destroy.bind(socket), 10)
    })

    server.listen(() => {
      const addr = server.address()
      const session = http2.connect('http://localhost:' + addr.port, {
        createConnection: (authority) => {
          return (socket = net.connect(authority.port, authority.hostname))
        }
      })

      const request = session.request({
        ':method': 'POST',
        ':path': '/',
        'content-length': '50'
      })

      request.write('testing...')
    })
  })
})

const http2close = (server, session, callback) => {
  const onSessionClose = () => {
    server.close(() => callback())
  }

  if (typeof session.close === 'function') {
    session.close(onSessionClose)
  } else {
    session.shutdown(onSessionClose)
  }
}


test/index.js
const assert = require('assert')
const fs = require('fs')
const getRawBody = require('..')
const path = require('path')

const Buffer = require('xprezzo-buffer').Buffer
const EventEmitter = require('events').EventEmitter
const Promise = global.Promise || require('bluebird')
const Readable = require('readable-stream').Readable

const file = path.join(__dirname, 'index.js')
const length = fs.statSync(file).size
const string = fs.readFileSync(file, 'utf8')

// Add Promise to mocha's global list
// eslint-disable-next-line no-self-assign
global.Promise = global.Promise

describe('Raw Body', () => {
  it('should work without any options', (done) => {
    getRawBody(createStream(), (err, buf) => {
      assert.ifError(err)
      checkBuffer(buf)
      done()
    })
  })

  it('should work with `true` as an option', (done) => {
    getRawBody(createStream(), true, (err, buf) => {
      assert.ifError(err)
      assert.strictEqual(typeof buf, 'string')
      done()
    })
  })

  it('should error for bad callback', () => {
    assert.throws(() => {
      getRawBody(createStream(), true, 'silly')
    }, /argument callback.*function/)
  })

  it('should work with length', (done) => {
    getRawBody(createStream(), {
      length: length
    }, (err, buf) => {
      assert.ifError(err)
      checkBuffer(buf)
      done()
    })
  })

  it('should work when length=0', (done) => {
    const stream = new EventEmitter()

    getRawBody(stream, {
      length: 0,
      encoding: true
    }, (err, str) => {
      assert.ifError(err)
      assert.strictEqual(str, '')
      done()
    })

    process.nextTick(() => {
      stream.emit('end')
    })
  })

  it('should work with limit', (done) => {
    getRawBody(createStream(), {
      limit: length + 1
    }, (err, buf) => {
      assert.ifError(err)
      checkBuffer(buf)
      done()
    })
  })

  it('should work with limit as a string', (done) => {
    getRawBody(createStream(), {
      limit: '1gb'
    }, (err, buf) => {
      assert.ifError(err)
      checkBuffer(buf)
      done()
    })
  })

  it('should work with limit and length', (done) => {
    getRawBody(createStream(), {
      length: length,
      limit: length + 1
    }, (err, buf) => {
      assert.ifError(err)
      checkBuffer(buf)
      done()
    })
  })

  it('should check options for limit and length', (done) => {
    getRawBody(createStream(), {
      length: length,
      limit: length - 1
    }, (err, buf) => {
      assert.strictEqual(err.status, 413)
      assert.strictEqual(err.statusCode, 413)
      assert.strictEqual(err.expected, length)
      assert.strictEqual(err.length, length)
      assert.strictEqual(err.limit, length - 1)
      assert.strictEqual(err.type, 'entity.too.large')
      assert.strictEqual(err.message, 'request entity too large')
      done()
    })
  })

  it('should work with an empty stream', (done) => {
    const stream = new Readable()
    stream.push(null)

    getRawBody(stream, {
      length: 0,
      limit: 1
    }, (err, buf) => {
      assert.ifError(err)
      assert.strictEqual(buf.length, 0)
      done()
    })

    stream.emit('end')
  })

  it('should throw on empty string and incorrect length', (done) => {
    const stream = new Readable()
    stream.push(null)

    getRawBody(stream, {
      length: 1,
      limit: 2
    }, (err, buf) => {
      assert.strictEqual(err.status, 400)
      done()
    })

    stream.emit('end')
  })

  it('should throw if length > limit', (done) => {
    getRawBody(createStream(), {
      limit: length - 1
    }, (err, buf) => {
      assert.strictEqual(err.status, 413)
      done()
    })
  })

  it('should throw if incorrect length supplied', (done) => {
    getRawBody(createStream(), {
      length: length - 1
    }, (err, buf) => {
      assert.strictEqual(err.status, 400)
      done()
    })
  })

  it('should work with if length is null', (done) => {
    getRawBody(createStream(), {
      length: null,
      limit: length + 1
    }, (err, buf) => {
      assert.ifError(err)
      checkBuffer(buf)
      done()
    })
  })

  it('should work with {"test":"å"}', (done) => {
    // https://github.com/visionmedia/express/issues/1816

    const stream = new Readable()
    stream.push('{"test":"å"}')
    stream.push(null)

    getRawBody(stream, {
      length: 13
    }, (err, buf) => {
      if (err) return done(err)
      assert.ok(buf)
      assert.strictEqual(buf.length, 13)
      done()
    })
  })

  it('should throw if stream encoding is set', (done) => {
    const stream = new Readable()
    stream.push('akl;sdjfklajsdfkljasdf')
    stream.push(null)
    stream.setEncoding('utf8')

    getRawBody(stream, (err, buf) => {
      assert.strictEqual(err.status, 500)
      done()
    })
  })

  it('should throw when given an invalid encoding', (done) => {
    const stream = new Readable()
    stream.push('akl;sdjfklajsdfkljasdf')
    stream.push(null)

    getRawBody(stream, 'akljsdflkajsdf', (err) => {
      assert.ok(err)
      assert.strictEqual(err.message, 'specified encoding unsupported')
      assert.strictEqual(err.status, 415)
      assert.strictEqual(err.type, 'encoding.unsupported')
      done()
    })
  })

  describe('with global Promise', () => {
    before(() => {
      global.Promise = Promise
    })

    after(() => {
      global.Promise = undefined
    })

    it('should work as a promise', () => {
      return getRawBody(createStream())
        .then(checkBuffer)
    })

    it('should work as a promise when length > limit', () => {
      return getRawBody(createStream(), {
        length: length,
        limit: length - 1
      }).then(throwExpectedError, (err) => {
        assert.strictEqual(err.status, 413)
      })
    })
  })

  describe('without global Promise', () => {
    before(() => {
      global.Promise = undefined
    })

    after(() => {
      global.Promise = Promise
    })

    it('should error without callback', () => {
      assert.throws(() => {
        getRawBody(createStream())
      }, /argument callback.*required/)
    })

    it('should work with callback as second argument', (done) => {
      getRawBody(createStream(), (err, buf) => {
        assert.ifError(err)
        checkBuffer(buf)
        done()
      })
    })

    it('should work with callback as third argument', (done) => {
      getRawBody(createStream(), true, (err, str) => {
        assert.ifError(err)
        checkString(str)
        done()
      })
    })
  })

  describe('when an encoding is set', () => {
    it('should return a string', (done) => {
      getRawBody(createStream(), {
        encoding: 'utf-8'
      }, (err, str) => {
        assert.ifError(err)
        assert.strictEqual(str, string)
        done()
      })
    })

    it('should handle encoding true as utf-8', (done) => {
      getRawBody(createStream(), {
        encoding: true
      }, (err, str) => {
        assert.ifError(err)
        assert.strictEqual(str, string)
        done()
      })
    })

    it('should handle encoding as options string', (done) => {
      getRawBody(createStream(), 'utf-8', (err, str) => {
        assert.ifError(err)
        assert.strictEqual(str, string)
        done()
      })
    })

    it('should decode codepage string', (done) => {
      const stream = createStream(Buffer.from('bf43f36d6f20657374e1733f', 'hex'))
      const string = '¿Cómo estás?'
      getRawBody(stream, 'iso-8859-1', (err, str) => {
        assert.ifError(err)
        assert.strictEqual(str, string)
        done()
      })
    })

    it('should decode UTF-8 string', (done) => {
      const stream = createStream(Buffer.from('c2bf43c3b36d6f20657374c3a1733f', 'hex'))
      const string = '¿Cómo estás?'
      getRawBody(stream, 'utf-8', (err, str) => {
        assert.ifError(err)
        assert.strictEqual(str, string)
        done()
      })
    })

    it('should decode UTF-16 string (LE BOM)', (done) => {
      // BOM makes this LE
      const stream = createStream(Buffer.from('fffebf004300f3006d006f002000650073007400e10073003f00', 'hex'))
      const string = '¿Cómo estás?'
      getRawBody(stream, 'utf-16', (err, str) => {
        assert.ifError(err)
        assert.strictEqual(str, string)
        done()
      })
    })

    it('should decode UTF-16 string (BE BOM)', (done) => {
      // BOM makes this BE
      const stream = createStream(Buffer.from('feff00bf004300f3006d006f002000650073007400e10073003f', 'hex'))
      const string = '¿Cómo estás?'
      getRawBody(stream, 'utf-16', (err, str) => {
        assert.ifError(err)
        assert.strictEqual(str, string)
        done()
      })
    })

    it('should decode UTF-16LE string', (done) => {
      // UTF-16LE is different from UTF-16 due to BOM behavior
      const stream = createStream(Buffer.from('bf004300f3006d006f002000650073007400e10073003f00', 'hex'))
      const string = '¿Cómo estás?'
      getRawBody(stream, 'utf-16le', (err, str) => {
        assert.ifError(err)
        assert.strictEqual(str, string)
        done()
      })
    })

    it('should correctly calculate the expected length', (done) => {
      const stream = createStream(Buffer.from('{"test":"å"}'))

      getRawBody(stream, {
        encoding: 'utf-8',
        length: 13
      }, done)
    })
  })

  it('should work on streams1 stream', (done) => {
    const stream = new EventEmitter()

    getRawBody(stream, {
      encoding: true,
      length: 19
    }, (err, value) => {
      assert.ifError(err)
      assert.strictEqual(value, 'foobar,foobaz,yay!!')
      done()
    })

    process.nextTick(() => {
      stream.emit('data', 'foobar,')
      stream.emit('data', 'foobaz,')
      stream.emit('data', 'yay!!')
      stream.emit('end')
    })
  })
})

const checkBuffer = (buf) => {
  assert.ok(Buffer.isBuffer(buf))
  assert.strictEqual(buf.length, length)
  assert.strictEqual(buf.toString('utf8'), string)
}

const checkString = (str) => {
  assert.ok(typeof str === 'string')
  assert.strictEqual(str, string)
}

const createStream = (buf) => {
  if (!buf) return fs.createReadStream(file)

  const stream = new Readable()
  stream._read = () => {
    stream.push(buf)
    stream.push(null)
  }

  return stream
}

const throwExpectedError = () => {
  throw new Error('expected error')
}


index.js
/**
 * xprezzo-raw-body
 * Copyright(c) 2022 Cloudgen Wong <cloudgen.wong@gmail.com>
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */

const bytes = require('xprezzo-bytes')
const HttpRawStream = require('./lib/HttpRawStream')
const createError = require('xprezzo-http-errors')
const iconv = require('xprezzo-iconv')
const onFinished = require('xprezzo-on-finished')
const zlib = require('zlib')

/**
  * Module exports.
  * Wrap and protect the HttpRawStream
  *
  * @param {object} stream
  * @param {object|string|function} [options]
  * @param {function} [callback]
  * @public
  */
const getBody = module.exports = (stream, options, callback) => {
  let done = callback
  let opts = options || {}

  if (options === true || typeof options === 'string') {
    // short cut for encoding
    opts = {
      encoding: options
    }
  } else if (typeof options === 'function') {
    done = options
    opts = {}
  }

  // validate callback is a function, if provided
  if (done !== undefined && typeof done !== 'function') {
    throw new TypeError('argument callback must be a function')
  }

  // require the callback without promises
  if (!done && !global.Promise) {
    throw new TypeError('argument callback is required')
  }

  // get encoding
  const encoding = opts.encoding !== true
    ? opts.encoding
    : 'utf-8'

  // convert the limit to an integer
  const limit = bytes.parse(opts.limit)

  // convert the expected length to an integer
  const length = opts.length != null && !isNaN(opts.length)
    ? parseInt(opts.length, 10)
    : null

  if (done) {
    // classic callback style
    // new HttpRawStream(stream, encoding, length, limit, callback);
    return new HttpRawStream({
      stream: stream,
      encoding: encoding,
      length: length,
      limit: limit,
      callback: done
    })
  }

  return new Promise((resolve, reject) => {
    /* eslint-disable no-new */
    new HttpRawStream({
      stream: stream,
      encoding: encoding,
      length: length,
      limit: limit,
      callback: (err, buf) => {
        if (err) return reject(err)
        resolve(buf)
      }
    })
  })
}

/**
 * Module exports.
 *
 * Reader
 *
 * Read a request into a buffer and parse.
 *
 * @param {object} req
 * @param {object} res
 * @param {function} next
 * @param {function} parse
 * @param {function} debug
 * @param {object} options
 * @private
 */
module.exports.Reader = (req, res, next, parse, debug, options) => {
  let length
  const opts = options
  let stream

  // flag as parsed
  req._body = true

  // read options
  const encoding = opts.encoding !== null
    ? opts.encoding
    : null
  const verify = opts.verify
  debug('typeof verify = ' + typeof verify)

  try {
    // get the content stream
    stream = contentstream(req, debug, opts.inflate)
    length = stream.length
    debug('length = ' + length)
    stream.length = undefined
  } catch (err) {
    return next(err)
  }

  // set xprezzo-raw-body options
  opts.length = length
  opts.encoding = verify
    ? null
    : encoding
  debug('encoding = ' + encoding)
  debug('encoding exists=' + iconv.encodingExists(encoding))

  // assert charset is supported
  if (opts.encoding === null && encoding !== null && !iconv.encodingExists(encoding)) {
    return next(createError(415, 'unsupported charset "' + encoding.toUpperCase() + '"', {
      charset: encoding.toLowerCase(),
      type: 'charset.unsupported'
    }))
  }

  // read body
  debug('read body')
  getBody(stream, opts, (error, body) => {
    if (error) {
      let _error

      if (error.type === 'encoding.unsupported') {
        // echo back charset
        _error = createError(415, 'unsupported charset "' + encoding.toUpperCase() + '"', {
          charset: encoding.toLowerCase(),
          type: 'charset.unsupported'
        })
      } else {
        // set status code on error
        _error = createError(400, error)
      }

      // read off entire request
      stream.resume()
      onFinished(req, function onfinished () {
        next(createError(400, _error))
      })
      return
    }

    // verify
    if (verify) {
      try {
        debug('verify body')
        verify(req, res, body, encoding)
      } catch (err) {
        next(createError(403, err, {
          body: body,
          type: err.type || 'entity.verify.failed'
        }))
        return
      }
    }

    // parse
    let str = body
    try {
      debug('parse body')
      str = typeof body !== 'string' && encoding !== null
        ? iconv.decode(body, encoding)
        : body
      req.body = parse(str)
    } catch (err) {
      next(createError(400, err, {
        body: str,
        type: err.type || 'entity.parse.failed'
      }))
      return
    }
    next()
  })
}

/**
 * Get the content stream of the request.
 *
 * @param {object} req
 * @param {function} debug
 * @param {boolean} [inflate=true]
 * @return {object}
 * @api private
 */

const contentstream = (req, debug, inflate) => {
  const encoding = (req.headers['content-encoding'] || 'identity').toLowerCase()
  const length = req.headers['content-length']
  let stream

  debug('content-encoding "%s"', encoding)

  if (inflate === false && encoding !== 'identity') {
    throw createError(415, 'content encoding unsupported', {
      encoding: encoding,
      type: 'encoding.unsupported'
    })
  }

  switch (encoding) {
    case 'deflate':
      stream = zlib.createInflate()
      debug('inflate body')
      req.pipe(stream)
      break
    case 'gzip':
      stream = zlib.createGunzip()
      debug('gunzip body')
      req.pipe(stream)
      break
    case 'identity':
      stream = req
      stream.length = length
      break
    default:
      throw createError(415, 'unsupported content encoding "' + encoding + '"', {
        encoding: encoding,
        type: 'encoding.unsupported'
      })
  }

  return stream
}
/**
 * Module exports.
 *
 */
module.exports.bytes = bytes
module.exports.debug = require('xprezzo-debug')
module.exports.httpErrors = require('xprezzo-http-errors')
module.exports.iconv = iconv
module.exports.onFinished = onFinished


lib/HttpRawStream.js
/*!
 * xprezzo-raw-body
 * Copyright(c) 2020 Ben Ajenoui <info@seohero.io>
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */
const debug = require('xprezzo-debug')('xprezzo:HttpRawStream')
const createError = require('xprezzo-http-errors')
const iconv = require('xprezzo-iconv')
const unpipe = require('xprezzo-stream-unpipe')
const prop = new WeakMap()

class HttpRawStream {
  constructor (options) {
    const opts = options || {}
    opts.complete = false
    opts.sync = true
    opts.state = opts.stream._readableState
    opts.received = 0
    opts.decoder = ''
    this.limit2 = opts.limit
    this.cleanup = function () {}
    prop.set(this, opts)
    const entitySizeResult = checkEntitySize.call(this)
    if (entitySizeResult) return entitySizeResult
    const streamEncodingResult = checkStreamEncoding.call(this)
    if (streamEncodingResult) return streamEncodingResult
    const decoderResult = useDecoder.call(this, opts.encoding)
    if (decoderResult) return decoderResult
    setListeners.call(this)
  }
}

/**
 * Get the decoder for a given encoding.
 *
 * @param {string} encoding
 * @private
 */
const getDecoder = (encoding) => {
  if (!encoding) return null
  try {
    return iconv.getDecoder(encoding)
  } catch (e) {
    // error getting decoder
    if (!/^encoding\s+not\s+recognized:\s+/i.test(e.message)) throw e

    // the encoding was not found
    throw createError(415, 'specified encoding unsupported', {
      encoding: encoding,
      type: 'encoding.unsupported'
    })
  }
}

/**
 * Check the stream encoding before start
 *
 * @private
 */
// Avoid using arrow function expression to prevent incorrect "this"
function checkStreamEncoding () {
  const self = prop.get(this)
  // streams1: assert request encoding is buffer.
  // streams2+: assert the stream encoding is buffer.
  //   stream._decoder: streams1
  //   state.encoding: streams2
  //   state.decoder: streams2, specifically < 0.10.6
  if (self.stream._decoder || (self.state && (self.state.encoding || self.state.decoder))) {
    // developer error
    return final.call(this, createError(500, 'stream encoding should not be set', {
      type: 'stream.encoding.set'
    }))
  }
  return false
}

/**
 * Final call handler
 *
 * @private
 */
// Avoid using arrow function expression to prevent incorrect "this"
function final () {
  const self = prop.get(this)
  const that = this
  const args = new Array(arguments.length)
  // copy arguments
  for (let i = 0; i < args.length; i++) {
    args[i] = arguments[i]
  }
  // mark complete
  self.complete = true
  prop.set(this, self)
  const invokeCallback = () => {
    that.cleanup()
    if (args[0]) {
      debug('callBack')
      // halt the stream on error
      halt.call(that)
    }
    self.callback.apply(null, args)
  }
  if (self.sync) {
    process.nextTick(invokeCallback)
  } else {
    invokeCallback()
  }
  return this
}

/**
 * halt the stream
 *
 * @private
 */
// Avoid using arrow function expression to prevent incorrect "this"
function halt () {
  const self = prop.get(this)
  // unpipe everything from the stream
  unpipe(self.stream)
  // pause stream
  if (typeof self.stream.pause === 'function') {
    self.stream.pause()
  }
}

/**
 * Check if request entity is too large
 *
 * @private
 */
// Avoid using arrow function expression to prevent incorrect "this"
function checkEntitySize () {
  const self = prop.get(this)
  // check the length and limit options.
  // note: we intentionally leave the stream paused,
  // so users should handle the stream themselves.
  if (self.limit !== null && self.length !== null && self.length > self.limit) {
    return final.call(this, createError(413, 'request entity too large', {
      expected: self.length,
      length: self.length,
      limit: self.limit,
      type: 'entity.too.large'
    }))
  }
  return false
}

/**
 * Apply Decoder
 *
 * @param {string} encoding
 *
 * @private
 */
// Avoid using arrow function expression to prevent incorrect "this"
function useDecoder (encoding) {
  const self = prop.get(this)
  try {
    self.decoder = getDecoder(encoding)
  } catch (err) {
    return final.call(this, err)
  }
  self.buffer = self.decoder ? '' : []
  prop.set(this, self)
  return false
}

/**
 * Set the listener for HttpRawStream
 *
 *
 * @private
 */
// Avoid using arrow function expression to prevent incorrect "this"
function setListeners () {
  const self = prop.get(this)
  debug('3. limit:' + self.limit)
  const that = this
  function cleanup () {
    self.buffer = null
    prop.set(that, self)
    self.stream.removeListener('aborted', onAborted)
    self.stream.removeListener('data', onData)
    self.stream.removeListener('end', onEnd)
    self.stream.removeListener('error', onEnd)
    self.stream.removeListener('close', cleanup)
  }
  this.cleanup = cleanup
  prop.set(this, self)
  // attach listeners
  self.stream.on('aborted', onAborted)
  self.stream.on('close', cleanup)
  self.stream.on('data', onData)
  self.stream.on('end', onEnd)
  self.stream.on('error', onEnd)
  // mark sync section complete
  self.sync = false
  prop.set(this, self)
  function onAborted () {
    HttpRawStreamOnAborted.call(that)
  }
  function onData (chunk) {
    HttpRawStreamOnData.call(that, chunk)
  }
  function onEnd (err) {
    HttpRawStreamOnEnd.call(that, err)
  }
  return this
}

/**
 * The On Abort Handler
 *
 * @private
 */
// Avoid using arrow function expression to prevent incorrect "this"
function HttpRawStreamOnAborted () {
  const self = prop.get(this)
  if (this.complete) return
  final.call(this, createError(400, 'request aborted', {
    code: 'ECONNABORTED',
    expected: self.length,
    length: self.length,
    received: self.received,
    type: 'request.aborted'
  }))
  return false
}

/**
 * The On Data Handler
 *
 * @private
 */
// Avoid using arrow function expression to prevent incorrect "this"
function HttpRawStreamOnData (chunk) {
  const self = prop.get(this)
  if (self.complete) return
  self.received += chunk.length
  prop.set(this, self)
  if (self.limit !== null && self.received > self.limit) {
    final.call(this, createError(413, 'request entity too large', {
      limit: self.limit,
      received: self.received,
      type: 'entity.too.large'
    }))
  } else if (self.decoder) {
    self.buffer += self.decoder.write(chunk)
    prop.set(this, self)
  } else {
    self.buffer.push(chunk)
    prop.set(this, self)
  }
  return false
}

/**
 * The On End Handler
 *
 * @private
 */
// Avoid using arrow function expression to prevent incorrect "this"
function HttpRawStreamOnEnd (err) {
  const self = prop.get(this)
  if (self.complete) return
  if (err) return final.call(this, err)
  if (self.length !== null && self.received !== self.length) {
    final.call(this, createError(400, 'request size did not match content length', {
      expected: self.length,
      length: self.length,
      received: self.received,
      type: 'request.size.invalid'
    }))
  } else {
    const string = self.decoder
      ? self.buffer + (self.decoder.end() || '')
      : Buffer.concat(self.buffer)
    final.call(this, null, string)
  }
  return false
}

module.exports = HttpRawStream
